EIPS/eip-4337.md

```diff
@@ -159,6 +158,72 @@ interface IAggregator {
 * **validateSignatures()** MUST validate the aggregated signature matches for all UserOperations in the array, and revert otherwise.
   This method is called on-chain by `handleOps()`
 
+#### Semi-abstracted Nonce Support
+
+In Ethereum protocol, the sequential transaction `nonce` value is used as a replay protection method as well as to
+determine the valid order of transaction being included in blocks.
+
+It also contributes to the transaction hash uniqueness, as a transaction by the same sender with the same
+nonce may not be included in the chain twice.
+
+However, requiring a single sequential `nonce` value is limiting the senders' ability to define their custom logic
+with regard to transaction ordering and replay protection.
+
+Instead of sequential `nonce` we implement a nonce mechanism that uses a single `uint256` nonce value in the `UserOperation`,
+but treats it as two values:
+
+* 192-bit "key"
+* 64-bit "sequence"
+
+These values are represented on-chain in the `EntryPoint` contract.
+We define the following method in the `EntryPoint` interface to expose these values:
+
+```solidity
+function getNonce(address sender, uint192 key) external view returns (uint256 nonce);
+```
+
+For each `key` the `sequence` is validated and incremented sequentially and monotonically by the `EntryPoint` for
```

Comment:
Is there any way that you want the contract to enforce that sequence can only go sequentially up, e..g throw erorrs?

Reply:
It's not enforced by the contract, but by EntryPoint itself.  See `_validateAndUpdateNonce` in NonceManager.sol and how it's used by EntryPoint.sol.  The nonce can only go sequentially for each key (which the user provides and the account can set restrictions on).  If it hasn't been incremented as expected, EntryPoint will throw `FailedOp(opIndex, "AA25 invalid account nonce")`.

Reply:
> Is there any way that you want the contract to enforce that sequence can only go sequentially up, e..g throw erorrs?

The account can enforce it into sequence-only, by requiring the key to be always zero 
(e.g. `require(userOp.nonce < type(uint64).max), "this account supports sequential nonces only")`)